ImageLounge/src/DkCore/DkImageStorage.cpp

```diff
@@ -280,6 +287,86 @@ QImage DkImage::rotateImage(const QImage &img, double angle)
     return imgR;
 }
 
+QImage transposeImage(const QImage &img)
+{
+    QImage imgIn = img;
+    if (img.depth() > 32) {
+        // Ensure the data is 32-bit aligned
+        imgIn = img.convertToFormat(QImage::Format_ARGB32);
+    }
+
+    QImage imgOut = QImage(imgIn.size().transposed(), imgIn.format());
+
+    const QRgb *dataIn = reinterpret_cast<const QRgb *>(imgIn.constBits());
+    QRgb *dataOut = reinterpret_cast<QRgb *>(imgOut.bits());
+
```

Comment:
I recommend using QImage::bytesPerline() for the row stride, or adding a check. This allows doing a zero-copy conversion from image buffers from another library (opencv or libraw, openGL, whatever, in a future-proof way.

Reply:
Not sure what you want to check for?

Following the docs, I ensure the image is 32-bit aligned, then cast the data pointer to `QRgb*`. The row stride is then the width.

![image](https://github.com/user-attachments/assets/a4d2701c-ed16-4e51-a0f3-c2b8603c916a)

Reply:
> The row stride is then the width.

I'm reading that it must be 32-bit aligned *at minimum*, so it could be 32,64,128, etc. In which case there can be unused bytes at the end of each row that must be considered. The QImage constructors that take a pointer and bytesperline argument might produce such an image, we might want to use these constructors at some point when working with large images, since they don't have to make a copy.

So I believe you must check if  bytesPerLine()==width()*bytesPerPixel() if you are wanting to treat the image  buffer as a continuous block.

Reply:
> it must be 32-bit aligned at minimum, so it could be 32,64,128

Yes, it must be 32-bit aligned at minimum, and 32-bpp images are also 32-bit aligned, otherwise they cannot be casted to `QRgb*` safely. So, we only need format conversion if we have 64-bit image formats, as they can be 32-bit or 64-bit aligned depending on platforms.

> The QImage constructors that take a pointer and bytesperline argument might produce such an image, we might want to use these constructors at some point when working with large images, since they don't have to make a copy.

I would probably avoid using these constructors because the resulting `QImage` will not own the memory. I try leveraging the copy-on-write (or "implicit data sharing" from the Qt docs) property of the `QImage` as much as possible. When there's a need, I can allocate a `QImage` from a known size, and access the memory directly with `bits()` or `constBits()`, so the memory stays managed by the `QImage`.